"switched" command runner

tdyas opened the pull request:
Introduce `SwitchedCommandRunner` which uses a predicate to choose between two command runners. Then use `SwitchedCommandRunner` to switch between the `docker::CommandRunner` and whatever other local command runner is configured.

This allows `docker::CommandRunner` to no longer know about any other command runners, thereby re-introducing better encapsulation to `docker::CommandRunner`.

Closes https://github.com/pantsbuild/pants/issues/16788

Eric-Arellano commented on src/rust/engine/src/context.rs:
```diff
@@ -258,6 +258,10 @@ impl Core {
         docker::ImagePullPolicy::OnlyIfLatestOrMissing,
       )?);
 
+      let runner =
+        SwitchedCommandRunner::new(docker_runner, runner, |req| req.docker_image.is_some());
+      let runner = Box::new(runner) as _;
```
Why not directly call `Box::new()` and do this in one line?

tdyas replied:
Beats me. Will change.

tdyas replied:
Fixed.

Eric-Arellano reviewed (approved):
Thanks!

stuhood reviewed (approved):
The `nailgun` command runner should be able to use this as well: https://github.com/pantsbuild/pants/blob/c7029014e7fb029223add37302c6247b2b648178/src/rust/engine/process_execution/src/nailgun/mod.rs#L79-L91

tdyas commented:
I switched this to using generic types instead of `Box<dyn CommandRunner>`. Then I added an impl of `CommandRunner` for `Box<T> where T: CommandRunner`.

tdyas commented:
> The `nailgun` command runner should be able to use this as well:

Will look into a follow-up.
